File: go/config/execution.go
package config

import (
	"fmt"
	"strconv"
	"strings"
)

// The execution providers, spelled as the shared contract spells them: the
// same five words in Python, Rust, Go and JS, so a benchmark row and a bug
// report name the same thing in every port.
const (
	ProviderAuto     = "auto"
	ProviderCPU      = "cpu"
	ProviderCUDA     = "cuda"
	ProviderCoreML   = "coreml"
	ProviderDirectML = "directml"
)

// ONNXProviders are the accepted values of ExecutionConfig.ONNXProvider.
var ONNXProviders = []string{
	ProviderAuto, ProviderCPU, ProviderCUDA, ProviderCoreML, ProviderDirectML,
}

// ProviderPreference is the order ProviderAuto tries the concrete providers,
// best first. CPU is last and always reachable, so auto always has an answer.
//
// auto prefers a provider only where a measurement says it is faster. CoreML
// is faster -- the split placement in the onnx package measures RTF 1.35-1.70
// on an M3 Pro against 0.85-1.02 for all-CPU -- and is still not here, for a
// reason that is not speed: compiling the renderer graphs costs about 146 s
// the first time on a machine and leaves 1.6 GB of cache behind, and a default
// may not spend either without being asked. DirectML has never been run by
// this project. Both stay selectable by name; neither is a default. CUDA leads
// until it is measured, and drops out the same way if it loses.
//
// This is what auto *prefers*, which is not the same question as what the
// library *offers*: see ConcreteProviders.
var ProviderPreference = []string{
	ProviderCUDA, ProviderCPU,
}

// ConcreteProviders are every provider that can actually run graphs, in the
// order an availability report lists them. ProviderAuto is absent: it is a
// request, not a provider.
//
// Kept apart from ProviderPreference because the two answer different
// questions. Probing only the preference list would make a provider that auto
// declines unreachable by name as well -- a caller asking for coreml would be
// told the library does not offer it, on a machine where it does.
var ConcreteProviders = []string{
	ProviderCUDA, ProviderCoreML, ProviderDirectML, ProviderCPU,
}

// ExecutionConfig is how the engine runs, never what it computes.
//
// Nothing here reaches the fingerprint. Two machines running one
// AlgorithmConfig on two providers agree on the recipe and may disagree in the
// last bits of the audio; that is the split loudkit.config draws between the
// two structs, and this port keeps it.
//
// This port carries the one field an ONNX binding can honour. Precision,
// attention kernel and the rest of Python's ExecutionConfig are properties of
// the torch path, and the graphs here are exported fp32.
type ExecutionConfig struct {
	// ONNXProvider is which execution provider runs the graphs, one of
	// ONNXProviders.
	//
	// The zero value "" reads as ProviderAuto, so ExecutionConfig{} and
	// ExecutionConfig{ONNXProvider: "auto"} are the same request. Go has no
	// unset field, and a struct literal must not mean something the named
	// default does not.
	ONNXProvider string
}

// DefaultExecution is what a caller who says nothing gets.
func DefaultExecution() ExecutionConfig {
	return ExecutionConfig{ONNXProvider: ProviderAuto}
}

// RequestedProvider is ONNXProvider with the zero value read as auto.
func (e ExecutionConfig) RequestedProvider() string {
	if e.ONNXProvider == "" {
		return ProviderAuto
	}
	return e.ONNXProvider
}

// Validate refuses a provider name outside ONNXProviders.
//
// Spelling is checked before anything is probed, because a typo that reached
// the resolver would come back as "this build does not offer metal" — which
// reads as a missing library and sends the caller off to install something
// that does not exist.
func (e ExecutionConfig) Validate() error {
	p := e.RequestedProvider()
	for _, known := range ONNXProviders {
		if p == known {
			return nil
		}
	}
	return fmt.Errorf("unknown onnx provider %q; expected one of %s",
		p, strings.Join(ONNXProviders, ", "))
}

// DescribeExecution is the execution line, mirroring Python's
// ExecutionConfig.describe.
//
// It takes the provider that was chosen, not the one that was asked for:
// "auto" is a question, and a benchmark row needs the answer.
//
// The two tokens are Python's, not this port's: "onnx" is the placement, where
// Python prints self.device, and "provider=" is the flag it appends beside it
// (config.py ExecutionConfig.describe). Rust prints the same pair. A grep for
// `provider=` across four ports' logs has to find four hits, so this spells it
// the way the others do rather than the shorter way.
func DescribeExecution(chosen string) string {
	return "exec[onnx provider=" + chosen + "]"
}

// Describe is the one-line algorithm summary, mirroring Python's
// AlgorithmConfig.describe field for field so the two ports' log lines can be
// diffed. It is a log line and not an identity: the fingerprint it opens with
// is the identity, and nothing in the rest of it is hashed.
func Describe(cfg AlgorithmConfig) string {
	g := cfg.Guidance
	if g != "single_path" {
		g = "cfg@" + pyFloat(cfg.GuidanceRate)
	}
	grid := "cosine"
	if len(cfg.EulerGrid) > 0 {
		grid = "explicit"
	}
	win := "ragged"
	if cfg.Window.StaticLength != nil {
		win = strconv.Itoa(*cfg.Window.StaticLength)
	}
	return fmt.Sprintf(
		"algo[%s] %s %s euler=%d(%s) temp=%s rep=%s min_p=%s sil=%d win=%s",
		Fingerprint(cfg), cfg.RecipeVersion, g, cfg.EulerSteps, grid,
		pyFloat(cfg.Sampling.Temperature),
		pyFloat(cfg.Sampling.RepetitionPenalty),
		pyFloat(cfg.Sampling.MinP),
		len(cfg.Sampling.SilenceTokenIds), win)
}

File: js/src/execution.ts
/**
 * Which onnxruntime execution provider runs the exported graphs.
 *
 * The execution layer, mirroring `loudkit.config.ExecutionConfig`: it decides
 * how fast, and — unlike the algorithm layer — a GPU provider is allowed to
 * move the last bits. Nothing here is part of the fingerprint.
 *
 * Before this, every port asked onnxruntime for nothing and got the CPU
 * provider, so the published benchmark figures described the torch path while
 * a JS caller silently ran at about real time.
 */

import { ort } from "./ort.js";

/**
 * The accepted spellings. These five words are the cross-language contract:
 * the Python, Rust and Go engines take the same names for the same concept,
 * because a port that spells one of them differently is indistinguishable
 * from a port that does something different. onnxruntime's own backend names
 * (`dml`, `CUDAExecutionProvider`) never leave this module.
 */
export const ONNX_PROVIDERS = ["auto", "cpu", "cuda", "coreml", "directml"] as const;

export type ONNXProvider = (typeof ONNX_PROVIDERS)[number];

/** A provider that names hardware — what `"auto"` resolves to. */
export type ResolvedONNXProvider = Exclude<ONNXProvider, "auto">;

/**
 * How the graphs run. Not the text, the voice or the seed, and not the
 * algorithm: two engines that disagree here are still computing the same
 * thing, up to the provider's own arithmetic.
 */
export interface ExecutionOptions {
  /**
   * Execution provider for every ONNX graph. `"auto"` — the default — takes
   * the best one this build and machine offer. An explicit value that is not
   * available is an error, never a quiet downgrade to CPU: a benchmark row
   * that says `cuda` and ran on CPU is the failure this option exists to
   * prevent.
   */
  onnxProvider?: ONNXProvider;
}

/**
 * The order `"auto"` searches. Not the same list as {@link ONNX_PROVIDERS},
 * which is a vocabulary and carries no preference.
 */
// auto prefers a provider only where a measurement says it is faster.
// CoreML EP measured 0.66x against the CPU provider's 1.36x on an M3 Pro,
// cost 25s of session load against 2s, and moved the token stream at index
// 41 (docs/benchmarks.md). DirectML has never been run by this project.
// Both stay selectable by name; neither is a default. CUDA leads until it
// is measured, and drops out the same way if it loses.
const REPORT_ORDER: readonly ResolvedONNXProvider[] = ["cuda", "coreml", "directml", "cpu"];

// What `auto` is willing to choose, which is narrower than what the build
// offers. Availability answers "can this run", preference answers "should it
// be the default", and they are not the same question.
const AUTO_PREFERENCE: readonly ResolvedONNXProvider[] = ["cuda", "cpu"];

/** loudkit's spelling -> the backend name onnxruntime-node reports and accepts. */
const ORT_BACKEND: Record<ResolvedONNXProvider, string> = {
  cpu: "cpu",
  cuda: "cuda",
  coreml: "coreml",
  directml: "dml",
};

const BY_ORT_BACKEND = new Map<string, ResolvedONNXProvider>(
  (Object.keys(ORT_BACKEND) as ResolvedONNXProvider[]).map((p) => [ORT_BACKEND[p], p])
);

/**
 * Where each provider comes from, for the error a caller who asked for a
 * missing one reads.
 *
 * onnxruntime-node ships one prebuilt binary per platform and arch, and which
 * providers that binary carries is fixed at build time — there is no separate
 * npm package to install for CUDA or DirectML, so the honest advice is the
 * platform matrix plus "build from source". CUDA is the one exception worth
 * naming: the binding carries it on linux/x64, but the EP's own shared
 * libraries are downloaded by the package's postinstall step, which
 * `--onnxruntime-node-install=skip` turns off.
 */
const PROVENANCE: Record<ResolvedONNXProvider, string> = {
  cpu:
    "every onnxruntime-node build carries the CPU provider, so a build that does not " +
    "is damaged — reinstall onnxruntime-node",
  cuda:
    "CUDA is in the linux/x64 prebuilt binary only (CUDA 12), and the EP's libraries are " +
    "fetched by the package's postinstall step, which `--onnxruntime-node-install=skip` " +
    "disables",
  coreml: "CoreML is in the darwin/x64 and darwin/arm64 prebuilt binaries only",
  directml: "DirectML is in the win32/x64 and win32/arm64 prebuilt binaries only",
};

/**
 * The providers this onnxruntime-node build offers, in loudkit's spelling.
 *
 * The build is asked rather than guessed from `process.platform`: the platform
 * says which prebuilt binary was downloaded, not what a locally built binding
 * was compiled with. Backends onnxruntime reports that loudkit has no name for
 * (`webgpu`, `tensorrt`) are dropped — a name outside the shared vocabulary
 * cannot be asked for, so listing it would only make the error longer.
 *
 * Takes the backend list as an argument so the resolution and its error can be
 * tested for hardware this machine does not have.
 */
export function availableProviders(
  backends: ReadonlyArray<{ name: string }> = ort.listSupportedBackends()
): ResolvedONNXProvider[] {
  const seen = new Set<ResolvedONNXProvider>();
  for (const b of backends) {
    const provider = BY_ORT_BACKEND.get(b.name);
    if (provider !== undefined) seen.add(provider);
  }
  // CoreML is dropped even where the binary carries it. This port refuses it
  // (see COREML_REFUSED), and a list that offered a provider the next call
  // rejects would send the reader looking for the wrong fault.
  return REPORT_ORDER.filter((p) => p !== "coreml" && seen.has(p));
}

/**
 * Why this port refuses CoreML, in the words the caller sees.
 *
 * CoreML is worth having only when the compiled models are cached: compiling
 * the renderer graphs takes about two minutes, and the other four ports pay
 * that once per machine by naming a cache directory. Node cannot name one. The
 * native addon reads `coreMlFlags` and nothing else -- its symbol table holds
 * no `ModelFormat` and no `ModelCacheDirectory` -- and `extra` session-config
 * entries are ignored, so ONNX Runtime's own EPContext caching is unreachable
 * too (both the nested and the flat key form were tried: no context file is
 * written and no error is raised).
 *
 * So every process would pay the compile again. Measured here on an M3 Pro:
 * 116.7 s to open the graphs against 2.1 s on CPU, for a renderer that is
 * faster only after those two minutes are spent. Offering that as a choice
 * would be offering a trap.
 */
const COREML_REFUSED =
  'onnxProvider "coreml" is not offered by this port. onnxruntime-node cannot ' +
  "cache compiled CoreML models -- its addon reads only coreMlFlags, and " +
  "session-config entries are ignored -- so every process pays the roughly " +
  "two-minute compile again: 116.7 s to open the graphs against 2.1 s on cpu. " +
  'Use "cpu" here. For CoreML on Apple hardware use the Swift package, or the ' +
  "Python, Rust or Go ports, which can name a model cache directory.";

/**
 * The providers to try, in order, for a requested value.
 *
 * One element for an explicit request, because an explicit request has no
 * fallback. For `"auto"` the whole available preference order, so that a
 * provider the build offers but the machine cannot actually run — a CUDA
 * binding whose EP libraries were never downloaded — costs a warning rather
 * than a failed load. Falling back is `"auto"`'s job and only `"auto"`'s; the
 * fallback itself lives in `openSessions`, which is where a session actually
 * gets opened.
 *
 * Throws for a value outside the vocabulary and for an explicit value this
 * build does not offer.
 */
export function resolveOnnxProvider(
  requested: ONNXProvider | undefined = "auto",
  available: readonly ResolvedONNXProvider[] = availableProviders()
): ResolvedONNXProvider[] {
  if (!(ONNX_PROVIDERS as readonly string[]).includes(requested)) {
    throw new Error(
      `onnxProvider must be one of ${ONNX_PROVIDERS.map((p) => `"${p}"`).join(", ")}, ` +
        `got ${JSON.stringify(requested)}`
    );
  }
  if (requested === "auto") {
    if (available.length === 0) {
      throw new Error(
        "onnxruntime-node reports no execution provider loudkit can use. Every build " +
          "carries the CPU provider, so this one is damaged — reinstall onnxruntime-node."
      );
    }
    return AUTO_PREFERENCE.filter((p) => available.includes(p));
  }
  if (requested === "coreml") {
    throw new Error(COREML_REFUSED);
  }
  if (!available.includes(requested)) {
    throw new Error(unavailable(requested, available));
  }
  return [requested];
}

/**
 * Names the provider asked for, the ones this build offers, and how to get the
 * missing one. All three, because each one alone sends the reader somewhere
 * else: without the offered list they cannot tell a typo from a missing build,
 * and without the provenance they go looking for an npm package that does not
 * exist.
 */
function unavailable(
  requested: ResolvedONNXProvider,
  available: readonly ResolvedONNXProvider[]
): string {
  return (
    `onnxProvider "${requested}" is not available. This onnxruntime-node build offers: ` +
    `${available.join(", ")}. ${PROVENANCE[requested]}; this process is ` +
    `${process.platform}/${process.arch}. For a provider your platform's prebuilt binary ` +
    "does not carry, build onnxruntime-node from source. Pass \"auto\" to take the best " +
    "provider this build offers."
  );
}

/** The session option onnxruntime-node wants for a resolved provider. */
export function executionProviders(provider: ResolvedONNXProvider): string[] {
  return [ORT_BACKEND[provider]];
}

/**
 * The execution half of `Engine.describe()`, in the `exec[...]` shape Python
 * and Swift print.
 *
 * `provider=` is the resolved name, never `"auto"`: a benchmark row and a bug
 * report both have to say which provider ran, and "auto" says nothing.
 */
export function describeExecution(provider: ResolvedONNXProvider): string {
  return `exec[onnx provider=${provider} prec[all=fp32]]`;
}
